src/io/obf.rs

```diff
@@ -213,62 +220,82 @@ pub async fn obfuscate<M, SU, SH, ST, R, P>(
         ));
 
         #[cfg(feature = "debug")]
-        handles.push(store_and_drop_matrix(
-            b_star_level.clone(),
-            &dir_path,
-            &format!("b_star_{level}"),
-        ));
-
-        for num in 0..level_size {
-            let mut handles_per_level: Vec<tokio::task::JoinHandle<()>> = Vec::new();
-            #[cfg(feature = "bgm")]
-            {
-                player.play_music(format!("bgm/obf_bgm{}.mp3", (2 * level + num) % 3 + 2));
-            }
-
-            let s_i_bar = sampler_uniform.sample_uniform(&params, d, d, DistType::BitDist);
-            let s_i_num = s_i_bar.concat_diag(&[&one_identity]);
-
+        store_and_drop_matrix(b_star_level.clone(), &dir_path, &format!("b_star_{level}"));
+
+        let u_nums_level = &u_nums[level - 1];
+        // Compute each levels in parallel with batching of 4
+        let batch_size = 4;
+        let k_preimages: Vec<_> = (0..level_size)
+            .collect::<Vec<_>>()
+            .par_chunks(batch_size)
+            .flat_map(|batch| {
+                batch
+                    .iter()
+                    .map(|&num| {
```

Comment:
Is it correct that 1) `level_size/batch_size` threads are parallelized and 2) each thread processes `batch` preimages sequentially?

Reply:
yes correct, and processing preimages sequentially is intentional as store_and_drop logic already handling tasks in queue and cannot be spawn in parallel.